fence/man/fence_vmware_vmrun.8
.TH fence_vmware_vmrun 8

.SH NAME
fence_vmware_vmrun - Another I/O Fencing agent for VMware virtual machines

.SH SYNOPSIS
.B
fence_vmware_vmrun
[\fIOPTION\fR]...

.SH DESCRIPTION
fence_vmware_vmrun is an I/O Fencing agent which can be used with the VMware ESX,
VMware ESXi or VMware Server to fence virtual machines. It is based on different
idea than fence_vmware and don't need to make ssh connection to ESX server. This
is main advantage, because ESXi doesn't have remote ssh administration console.

Before you can use this agent, it must be installed vmrun command on every node
you want to make fencing. Vmrun command is available as part of VMware VIX API
SDK package, which is available from VMware web pages (not in official RHEL
repository!). After you have successfully installed VIX API, you should be able
to run vmrun command. This agent supports only vmrun from version 2.0.0 (VIX API
1.6.0).

fence_vmware_vmrun accepts options on the command line as well as from stdin.
Fenced sends parameters through stdin when it execs the agent. fence_vmware_vmrun
can be run by itself with command line options. This is useful for testing
and for turning outlets on or off from scripts.

.SH HOST NAME/IP ADDRESS SPECIFICATION
Host name may be entered in many ways and this string is forwarded directly to
vmrun command. Basic usage is protocol://host_name:[port]/sdk. If you want
secure (https) connection, just use https as protocol, otherwise use http
(this is reason, why agent doesn't have any secure switch). Port isn't
mandatory, but you must use it, if you try use this agent against VMware
server 2.0.0 with default options, where https console runs on port 8333 and
http on 8222. In case of ESX/ESXi, you don't need to specify port, because
default configuration runs https on standard (443) port.

Don't forget to add /sdk suffix, otherwise agent will not work!

.SH OPTIONS
.TP
\fB-a\fP \fIIPaddress\fR
IP address or hostname of ESX/ESXi host machine. See (see
.SM
.B "HOST NAME/IP ADDRESS SPECIFICATION"
above).
.TP
\fB-h\fP
Print out a help message describing available options, then exit.
.TP
\fB-l\fP \fIlogin\fR
Login name to VMware administration console.
.TP
\fB-o\fP \fIaction\fR
The action required.  This can be reboot (default), status, off, on, list
or monitor.
.TP
\fB-p\fP \fIpassword\fR
Password for login for VMware administration console.
.TP
\fB-S\fP \fIscript\fR
Script to run to retrieve password for VMware administration console.
.TP
\fB-n\fP \fIname\fR
Name of virtual machine to fence (ie.[datastore1] test/test.vmx) which
is returned by list command.
.TP
\fB-c\fP \fIcommand\fR
Location of vmrun command. Default is /usr/bin/vmrun, but if you are using
nonstandard location, you may find this switch useful.
.TP
\fB-d\fP \fItype\fR
Type of VMware product you are trying to connect. This is directly forwarded
to vmrun command as -T switch. Default is esx, which is internally alias
for server (2.0.x) , which use same VIX API as ESX and ESXi.
.TP
\fB-T\fP
Test only.  Answer NO to the confirmation prompt instead of YES.
.TP
\fB-v\fP
Verbose. Record session to stdout, or debug file if specified (see -D).
.TP
\fB-D\fP
Specifies file, where will be written debug messages from session.
.TP
\fB-V\fP
Print out a version message, then exit.

.SH STDIN PARAMETERS
.TP
\fIagent = < param >\fR
This option is used by fence_node(8) and is ignored by fence_vmware_vmrun.
.TP
\fIipaddr = < hostname | ip >\fR
IP address or hostname of ESX/ESXi host machine. See (see
.SM
.B "HOST NAME/IP ADDRESS SPECIFICATION"
above).
.TP
\fIlogin = < param >\fR
Login name to VMware administration console.
.TP
\fIoption = < param >\fR
The action required.  This can be reboot (default), status, off, on, list
or monitor.
.TP
\fIpasswd = < param >\fR
Password for login for VMware administration console.
.TP
\fIpasswd_script = < param >\fR
Script to run to retrieve password for VMware administration console.
.TP
\fIport = < param >\fR
Name of virtual machine to fence (ie.[datastore1] test/test.vmx) which
is returned by list command.
.TP
\fIvmrun_cmd = < param >\fR
Location of vmrun command. Default is /usr/bin/vmrun, but if you are using
nonstandard location, you may find this switch useful.
.TP
\fIhost_type = < param >\fR
Type of VMware product you are trying to connect. This is directly forwarded
to vmrun command as -T switch. Default is esx, which is internally alias
for server (2.0.x) , which use same VIX API as ESX and ESXi.
.TP
\fItest = < param >\fR
Test only.  Answer NO to the confirmation prompt instead of YES.
.TP
\fIverbose = < param >\fR
Verbose.  Record session to stdout, or debug file if specified (see debug).
.TP
\fIdebug = < param >\fR
Specifies file, where will be written debug messages from session.

.SH SEE ALSO
fence(8), fence_node(8), fence_vmware(8)


fence/agents/vmware_vmrun/fence_vmware_vmrun.py
#!/usr/bin/python

#
# The Following agent has been tested on:
# vmrun 2.0.0 build-116503 (from VMware Server 2.0) against:
# 	VMware ESX 3.5 (works correctly)
# 	VMware Server 2.0.0 (works correctly)
#	VMware ESXi 3.5 update 2 (works correctly)
# 	VMware Server 1.0.7 (doesn't work)
# Any older version of vmrun doesn't have support for ESX/ESXi
#

import sys, re, pexpect
sys.path.append("@FENCEAGENTSLIBDIR@")
from fencing import *

#BEGIN_VERSION_GENERATION
RELEASE_VERSION="VMware Agent using VIX API"
REDHAT_COPYRIGHT=""
BUILD_DATE=""
#END_VERSION_GENERATION

# Path to vmrun command
VMRUN_COMMAND="/usr/bin/vmrun"
# Default type for -T parameter of vmrun command (default is esx, can be changed with -c)
VMWARE_DEFAULT_TYPE="esx"
# Minimum required version of vmrun command
VMRUN_MINIMUM_REQUIRED_VERSION=2

# Return string with command and additional parameters (something like vmrun -h 'host'
def vmware_vix_prepare_command(options,add_login_params,additional_params):
	res=options["-c"]

	if (add_login_params):
		res+=" -h '%s' -u '%s' -p '%s' -T '%s' "%(options["-a"],options["-l"],options["-p"],options["-d"])

	if (additional_params!=""):
		res+=additional_params

	return res

# Log message if user set verbose option
def vmware_vix_log(options, message):
	if options["log"] >= LOG_MODE_VERBOSE:
		options["debug_fh"].write(message+"\n")

# Run vmrun command with timeout and parameters. Internaly uses vmware_vix_prepare_command. Returns string
# with output from vmrun command. If something fails (command not found, exit code is not 0), fail_usage
# function is called (and never return).
def vmware_vix_run_command(options,add_login_params,additional_params):
	command=vmware_vix_prepare_command(options,add_login_params,additional_params)

	try:
		vmware_vix_log(options,command)

		(res_output,res_code)=pexpect.run(command,POWER_TIMEOUT+SHELL_TIMEOUT+LOGIN_TIMEOUT,True)

		if (res_code==None):
			fail(EC_TIMED_OUT)
		if ((res_code!=0) and (add_login_params)):
			vmware_vix_log(options,res_output)
			fail_usage("vmrun returned %s"%(res_output))
		else:
			vmware_vix_log(options,res_output)

	except pexpect.ExceptionPexpect:
		fail_usage(("Bad command name %s. Make sure, that you installed\nvmrun command (VIX Api). "+\
		"If you have nonstandard installation location,\ntry use -c switch.")%(options["-c"]))

	return res_output

# Returns True, if user uses supported vmrun version (currently >=2.0.0) otherwise False.
def vmware_vix_is_supported_vmrun_version(options):
	vmware_help_str=vmware_vix_run_command(options,False,"")
	version_re=re.search("vmrun version (\d\.(\d[\.]*)*)",vmware_help_str.lower())
	if (version_re==None):
		    return False   # Looks like this "vmrun" is not real vmrun

	version_array=version_re.group(1).split(".")

	try:
		if (int(version_array[0])<VMRUN_MINIMUM_REQUIRED_VERSION):
			return False
	except Exception:
		return False

	return True

def get_outlets_status(conn, options):
	outlets={}

	running_machines=vmware_vix_run_command(options,True,"list")
	all_machines=vmware_vix_run_command(options,True,"listRegisteredVM")

	all_machines_array=all_machines.splitlines()[1:]
	running_machines_array=running_machines.splitlines()[1:]

	for machine in all_machines_array:
		if (machine!=""):
			outlets[machine]=("",((machine in running_machines_array) and "on" or "off"))

	return outlets

def get_power_status(conn,options):
	outlets=get_outlets_status(conn,options)

	if (not (options["-n"] in outlets)):
		fail_usage("Failed: You have to enter existing name of virtual machine!")
	else:
		return outlets[options["-n"]][1]

def set_power_status(conn, options):
	additional_params="%s '%s'"%((options["-o"]=="on" and "start" or "stop"),options["-n"])
	if (options["-o"]=="off"):
		additional_params+=" hard"

	vmware_vix_run_command(options,True,additional_params)

# Define new options
def vmware_vix_define_new_opts():
	all_opt["vmrun_cmd"]={
		"getopt":"c:",
		"help":"-c <command>   Name of vmrun command (default "+VMRUN_COMMAND+")",
		"order": 2}
	all_opt["host_type"]={
		"getopt":"d:",
		"help":"-d <type>      Type of VMware to connect (default "+VMWARE_DEFAULT_TYPE+")",
		"order": 2}

def main():
	device_opt = [  "help", "version", "agent", "quiet", "verbose", "debug",
			"action", "ipaddr", "login", "passwd", "passwd_script",
			"test", "port", "separator", "vmrun_cmd", "host_type" ]

	vmware_vix_define_new_opts()

	options = check_input(device_opt, process_input(device_opt))

	# Fence agent specific defaults
        if (not options.has_key("-c")):
		options["-c"]=VMRUN_COMMAND

	if (not options.has_key("-d")):
		options["-d"]=VMWARE_DEFAULT_TYPE

	# Test user vmrun command version
	if (not (vmware_vix_is_supported_vmrun_version(options))):
		fail_usage("Unsupported version of vmrun command! You must use at least version %d!"%(VMRUN_MINIMUM_REQUIRED_VERSION))

	# Operate the fencing device
	fence_action(None, options, set_power_status, get_power_status, get_outlets_status)

if __name__ == "__main__":
	main()
